File: android/app/src/main/java/com/pickonemeal/data/notifications/NotificationSettingsRepository.kt
package com.pickonemeal.data.notifications

import com.pickonemeal.data.api.SupabaseClientProvider
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class NotificationSettingsDto(
    @SerialName("leading_chef_picks_enabled") val leadingChefPicksEnabled: Boolean? = null,
    @SerialName("table_winner_enabled") val tableWinnerEnabled: Boolean? = null,
    @SerialName("fcm_token") val fcmToken: String? = null
)

data class NotificationSettings(
    val leadingChefPicksEnabled: Boolean,
    val tableWinnerEnabled: Boolean,
    val fcmToken: String?
)

object NotificationSettingsRepository {

    suspend fun fetch(userId: String): NotificationSettings {
        val client = SupabaseClientProvider.client

        return try {
            val dto = client
                .from("user_notification_settings")
                .select("leading_chef_picks_enabled,table_winner_enabled,fcm_token")
                .eq("user_id", userId)
                .maybeSingle()
                .decodeAs<NotificationSettingsDto?>()

            NotificationSettings(
                leadingChefPicksEnabled = dto?.leadingChefPicksEnabled ?: true,
                tableWinnerEnabled = dto?.tableWinnerEnabled ?: true,
                fcmToken = dto?.fcmToken
            )
        } catch (_: Throwable) {
            NotificationSettings(
                leadingChefPicksEnabled = true,
                tableWinnerEnabled = true,
                fcmToken = null
            )
        }
    }

    suspend fun upsert(
        userId: String,
        leadingChefPicksEnabled: Boolean,
        tableWinnerEnabled: Boolean
    ) {
        val client = SupabaseClientProvider.client
        client
            .from("user_notification_settings")
            .upsert(
                mapOf(
                    "user_id" to userId,
                    "leading_chef_picks_enabled" to leadingChefPicksEnabled,
                    "table_winner_enabled" to tableWinnerEnabled
                ),
                onConflict = "user_id"
            )
    }

    suspend fun updateFcmToken(userId: String, token: String) {
        val client = SupabaseClientProvider.client
        client
            .from("user_notification_settings")
            .upsert(
                mapOf(
                    "user_id" to userId,
                    "fcm_token" to token
                ),
                onConflict = "user_id"
            )
    }
}

File: ios/Data/Notifications/NotificationSettingsRepository.swift
import Foundation
import Supabase

struct NotificationSettings: Decodable {
    var leadingChefPicksEnabled: Bool
    var tableWinnerEnabled: Bool
    var apnsToken: String?
}

@MainActor
final class NotificationSettingsRepository {
    private let client = SupabaseClientProvider.shared.client

    func fetchSettings(for userId: String) async throws -> NotificationSettings {
        struct Row: Decodable {
            let leading_chef_picks_enabled: Bool?
            let table_winner_enabled: Bool?
            let apns_token: String?
        }

        let response = try await client
            .from("user_notification_settings")
            .select("leading_chef_picks_enabled, table_winner_enabled, apns_token")
            .eq(column: "user_id", value: userId)
            .maybeSingle()

        if let response {
            let row = try response.decoded(to: Row.self)
            return NotificationSettings(
                leadingChefPicksEnabled: row.leading_chef_picks_enabled ?? true,
                tableWinnerEnabled: row.table_winner_enabled ?? true,
                apnsToken: row.apns_token
            )
        } else {
            // Default: all notifications on
            return NotificationSettings(
                leadingChefPicksEnabled: true,
                tableWinnerEnabled: true,
                apnsToken: nil
            )
        }
    }

    func upsertSettings(
        for userId: String,
        leadingChefPicksEnabled: Bool,
        tableWinnerEnabled: Bool
    ) async throws {
        _ = try await client
            .from("user_notification_settings")
            .upsert(
                values: [
                    "user_id": userId,
                    "leading_chef_picks_enabled": leadingChefPicksEnabled,
                    "table_winner_enabled": tableWinnerEnabled
                ],
                onConflict: "user_id"
            )
            .execute()
    }

    func updateApnsToken(for userId: String, token: String) async throws {
        _ = try await client
            .from("user_notification_settings")
            .upsert(
                values: [
                    "user_id": userId,
                    "apns_token": token
                ],
                onConflict: "user_id"
            )
            .execute()
    }
}


